agreementbot/pattern_manager.go

```diff
@@ -128,8 +135,57 @@ func (pm *PatternManager) hasPattern(org string, pattern string) bool {
 	return false
 }
 
-// Given a list of org/pattern pairs that this agbot is supported to be serving, take that list and
-// convert it to map of maps (keyed by org and pattern name) to hold all the pattern metadata. This
+// copy the given map of served patterns
+func (pm *PatternManager) setServedPatterns(servedPatterns map[string]exchange.ServedPattern) {
+	pm.spMapLock.Lock()
+	defer pm.spMapLock.Unlock()
+
+	// copy the input map
+	pm.ServedPatterns = servedPatterns
+}
+
+// chek if the agbot serves the given pattern or not.
+func (pm *PatternManager) servePattern(pattern_org string, pattern string) bool {
+	for _, sp := range pm.ServedPatterns {
```

Comment:
I think you need the spMapLock here to protect against write while you are reading. There may be other places too.

Reply:
servePattern() is in the same same process with the setServedPatterns(). The only thing that runs in a different process is GetServedNodeOrgs() which has spMapLock already.  Thoug it is safe to add spMapLock in servePattern() and servePattern() just in case someday someone will cal them from a different process. So I will add spMapLock for them.